state/watcher.go

```diff
@@ -496,6 +497,185 @@ func watchApplicationRelations(backend modelBackend, applicationName string) Str
+	if err != nil {
+		return errors.Trace(err)
+	}
+	changeSet = initialSet.Values()
+	outCh = w.outCh
+
+	for {
+		select {
+		case <-w.tomb.Dying():
+			return tomb.ErrDying
+		case <-w.watcher.Dead():
+			return stateWatcherDeadError(w.watcher.Err())
+		case changes := <-docWatcher.Changes():
+			if len(changes) == 0 {
+				continue
+			}
+			for _, docID := range changes {
+				unprocessedDocs.Add(w.backend.docID(docID))
+			}
+
+			// Restart the timer if currently stopped.
+			if !timerArmed {
+				_ = timer.Reset(w.quiesceInterval)
+			}
+		case <-timer.Chan():
+			timerArmed = false
+			if len(unprocessedDocs) == 0 {
+				continue // nothing to process
+			}
+
+			changedIDs, err := w.processChanges(unprocessedDocs)
+			unprocessedDocs = make(set.Strings)
+			if err != nil {
+				return err
+			} else if changedIDs.IsEmpty() {
+				continue // nothing to report
+			}
+
+			if len(changeSet) == 0 {
+				changeSet = changedIDs.Values()
+				outCh = w.outCh
+			} else {
+				// Append new set of changes to the not yet consumed changeset
+				changeSet = append(changeSet, changedIDs.Values()...)
+			}
+		case outCh <- changeSet:
+			changeSet = changeSet[:0]
+			outCh = nil
+		}
+	}
+}
+
+func (w *modelMachineStartTimeWatcher) initialMachineSet() (set.Strings, error) {
+	coll, closer := w.db.GetCollection(machinesC)
+	defer closer()
+
+	// Select the fields we need from documents that are not referring to
+	// containers.
+	iter := coll.Find(
+		bson.D{
+			{"$or", []bson.D{
+				{{"containertype", ""}},
+				{{"containertype", bson.D{{"$exists", false}}}},
+			}},
+		},
+	).Select(bson.D{{"_id", 1}, {"life", 1}, {"agent-started-at", 1}}).Iter()
+
+	var (
+		doc modelMachineStartTimeFieldDoc
+		ids = make(set.Strings)
+	)
+	for iter.Next(&doc) {
+		id := w.backend.localID(doc.Id)
+		ids.Add(id)
+		if doc.Life != Dead {
+			w.seenDocs[id] = doc
+		}
+	}
+	return ids, iter.Close()
+}
+
+func (w *modelMachineStartTimeWatcher) processChanges(pendingDocs set.Strings) (set.Strings, error) {
+	coll, closer := w.db.GetCollection(machinesC)
+	defer closer()
+
+	// Select the fields we need from the changed documents that are not
+	// referring to containers.
+	iter := coll.Find(
+		bson.D{
+			{"_id", bson.D{{"$in", pendingDocs.Values()}}},
+			{"$or", []bson.D{
+				{{"containertype", ""}},
+				{{"containertype", bson.D{{"$exists", false}}}},
+			}},
+		},
+	).Select(bson.D{{"_id", 1}, {"life", 1}, {"agent-started-at", 1}}).Iter()
+
+	var (
+		doc modelMachineStartTimeFieldDoc
+		ids = make(set.Strings)
+	)
+	for iter.Next(&doc) {
+		id := w.backend.localID(doc.Id)
+		old, exists := w.seenDocs[id]
+		if !exists || old.Life != doc.Life || old.AgentStartedAt != doc.AgentStartedAt {
+			w.seenDocs[id] = doc
+			ids.Add(id)
+		}
+
+		// If the machine is now dead we won't see a change for it again
+		// and therefore can permanently remove its entry from docHash
+		if doc.Life == Dead {
+			delete(w.seenDocs, id)
+		}
+	}
```

Comment:
interesting that you delete Dead docs, but not docs that we looked for but didn't find.
Is it worth tracking that set, and then doing a set difference?
I'm not sure if we need to cleanup seenDocs particularly aggressively

Reply:
My assumption here is that if they don't appear dead they might pop up in the future so removing the unseen ones didn't make sense to me.

The decision to prune Dead ones has more to do about long-running models where you start and stop machines all the time; unless we clean them up we will end up with a (tiny in the grand scheme of things) leak.

Reply:
Machines go Dying => Dead => removed from the collection.
I don't think there is a guarantee that you will see the object in Dead before it gets reaped. (While it is Dying you are supposed to remove your references, such that it doesn't go Dead until nobody refers to it anymore).

If we want to prune Dead,then we should prune objects-not-found for the same reason.

Reply:
This is an interesting point... when the instance-poller gets triggered it will call `Refresh` on the machine and then check if it's dead to remove it from the list (see https://github.com/juju/juju/blob/develop/worker/instancepoller/worker.go#L239-L247). 

If the machine doc gets reaped before it gets to do that then the worker will exit with an error. Do you think I should also patch the instance-poller to treat `errors.NotFound` when refreshing the machine as equivalent to being dead?

If I make the above change, I can then do a set difference to remove gone IDs from the map **and** also add them to the ID list that is emitted by the watcher. Thoughts?

Reply:
I wish we had better names for 'uuid#id' vs just 'id', caues its quite confusing what is a localID and what is global, and which set/collection is using which types of ids.
However, if you've tested this well, I'm happy with it.